drivers/pwm/pwm_nrfx.c

```diff
@@ -0,0 +1,216 @@
+		}
+
+		pdata->top_value = period_cycles;
+		nrf_pwm_configure(pconfig->pwm.p_registers,
+				  pconfig->config.base_clock, pconfig->config.count_mode,
+				  pdata->top_value);
+	}
+
+	if (pulse_cycles > pdata->top_value) {
+		return -EINVAL;
+	}
+
+	pdata->current[pwm] = (pdata->current[pwm] & ~PWM_COUNTERTOP_COUNTERTOP_Msk) | pulse_cycles;
+
+	return 0;
+}
+
+static int pwm_nrfx_get_cycles_per_sec(struct device *dev, u32_t pwm,
+				       u64_t *cycles)
+{
+	const struct pwm_nrfx_config *pconfig = dev->config->config_info;
+
+	switch (pconfig->config.base_clock) {
+	case NRF_PWM_CLK_16MHz:
+		*cycles = 16ul * 1000ul * 1000ul;
+		break;
+	case NRF_PWM_CLK_8MHz:
+		*cycles = 8ul * 1000ul * 1000ul;
+		break;
+	case NRF_PWM_CLK_4MHz:
+		*cycles = 4ul * 1000ul * 1000ul;
+		break;
+	case NRF_PWM_CLK_2MHz:
+		*cycles = 2ul * 1000ul * 1000ul;
+		break;
+	case NRF_PWM_CLK_1MHz:
+		*cycles = 1ul * 1000ul * 1000ul;
+		break;
+	case NRF_PWM_CLK_500kHz:
+		*cycles = 500ul * 1000ul;
+		break;
+	case NRF_PWM_CLK_250kHz:
+		*cycles = 250ul * 1000ul;
+		break;
+	case NRF_PWM_CLK_125kHz:
+		*cycles = 125ul * 1000ul;
+		break;
+	}
+	return 0;
+}
+
+static const struct pwm_driver_api pwm_nrfx_drv_api_funcs = {
+	.pin_set = pwm_nrfx_pin_set,
+	.get_cycles_per_sec = pwm_nrfx_get_cycles_per_sec,
+};
+
+static int pwm_nrfx_init(struct device *dev)
+{
+	const struct pwm_nrfx_config *pconfig = dev->config->config_info;
+
+	nrfx_err_t result = nrfx_pwm_init(&pconfig->pwm, &pconfig->config,
+					  NULL);
+
+	if (result != NRFX_SUCCESS) {
+		SYS_LOG_ERR("Failed to initialize device: %s",
+			    dev->config->name);
+		return -EBUSY;
+	}
+
+	nrfx_pwm_simple_playback(&pconfig->pwm, &pconfig->seq, 1, NRFX_PWM_FLAG_LOOP);
+
+	return 0;
+}
+
+
+#define PWM_NRFX_OUTPUT_PIN(dev_idx, ch_idx)		     \
+	CONFIG_PWM_##dev_idx##_NRFX_CH##ch_idx##_PIN |	     \
+	(CONFIG_PWM_##dev_idx##_NRFX_CH##ch_idx##_INVERTED ? \
+	 NRFX_PWM_PIN_INVERTED : 0)
+
+#define PWM_NRFX_DEFAULT_VALUE(dev_idx, ch_idx)		     \
+	(CONFIG_PWM_##dev_idx##_NRFX_CH##ch_idx##_INVERTED ? \
+	 PWM_NRFX_CH_VALUE_INVERTED : PWM_NRFX_CH_VALUE_NORMAL)
+
+#define PWM_NRFX_DEVICE(idx)								     \
+static struct pwm_nrfx_data pwm_nrfx_##idx##_data = {					     \
+	.current = {									     \
+		PWM_NRFX_DEFAULT_VALUE(idx, 0),						     \
+		PWM_NRFX_DEFAULT_VALUE(idx, 1),						     \
+		PWM_NRFX_DEFAULT_VALUE(idx, 2),						     \
+		PWM_NRFX_DEFAULT_VALUE(idx, 3),						     \
+	},										     \
+	.top_value = NRFX_PWM_DEFAULT_CONFIG_TOP_VALUE					     \
+};											     \
+	static const struct pwm_nrfx_config pwm_nrfx_##idx##_config = {			     \
+	.pwm = NRFX_PWM_INSTANCE(idx),							     \
+	.config = {									     \
+		.output_pins = {							     \
+			PWM_NRFX_OUTPUT_PIN(idx, 0),					     \
+			PWM_NRFX_OUTPUT_PIN(idx, 1),					     \
+			PWM_NRFX_OUTPUT_PIN(idx, 2),					     \
+			PWM_NRFX_OUTPUT_PIN(idx, 3),					     \
+		},									     \
+		.irq_priority = NRFX_PWM_DEFAULT_CONFIG_IRQ_PRIORITY,			     \
```

Comment:
This can be skipped, to not create confusion - interrupts are not used by this driver, and even if they were, the priority would need to be set via IRQ_CONNECT (this field is simply ignored).